gnovm/pkg/gnolang/store.go

```diff
@@ -107,16 +109,29 @@ func (ds *defaultStore) SetPackageGetter(pg PackageGetter) {
 	ds.pkgGetter = pg
 }
 
+type importerStore struct {
+	*defaultStore
+	importChain []string
+}
+
+func (is importerStore) GetPackage(pkgPath string, isImport bool) *PackageValue {
+	if !isImport {
+		// if not an import, match behaviour to the defaultStore
+		return is.defaultStore.GetPackage(pkgPath, isImport)
+	}
+	// it is an import -- detect cyclic imports
+	if slices.Contains(is.importChain, pkgPath) {
+		panic(fmt.Sprintf("import cycle detected: %q (through %v)", pkgPath, is.importChain))
+	}
+	return is.getPackage(pkgPath, is)
+}
+
 // Gets package from cache, or loads it from baseStore, or gets it from package getter.
 func (ds *defaultStore) GetPackage(pkgPath string, isImport bool) *PackageValue {
-	// detect circular imports
-	if isImport {
-		if _, exists := ds.current[pkgPath]; exists {
-			panic(fmt.Sprintf("import cycle detected: %q", pkgPath))
-		}
-		ds.current[pkgPath] = struct{}{}
-		defer delete(ds.current, pkgPath)
-	}
+	return ds.getPackage(pkgPath, importerStore{})
+}
+
+func (ds *defaultStore) getPackage(pkgPath string, impStore importerStore) *PackageValue {
```

Comment:
Why would we want to get a package by chaining another store passed in from other places?

Reply:
https://github.com/gnolang/gno/pull/1702/files/e20ebbe0baf93f0f279f2d575bd29e1230c0b1c4#r1571424655

Because the current method is thread unsafe and will fail if the Store is being used in two places at the same time.